openwisp_monitoring/device/apps.py

```diff
@@ -109,27 +111,79 @@ def is_working_changed_receiver(cls, instance, is_working, **kwargs):
         device_monitoring = device.monitoring
         initial_status = device_monitoring.status
         status = 'ok' if is_working else 'problem'
+        notification_opts = dict(sender=instance, target=device)
         if not is_working:
-            # Create a related notification explaining why it's not working
-            desc = instance.failure_reason
-            opts = dict(
-                sender=device, level='warning', verb='not working', description=desc
-            )
+            # Select appropriate notification type for reason of not working
+            if 'upgrade' in instance.failure_reason:
```

Comment:
This is too fragile. It will stop working and be forgotten if the behavior in openwisp-controller or openwisp-firmware-upgrader changes.

Please let's not do it.